repo/blob/azure/azure_pit.go

```diff
@@ -0,0 +1,203 @@
+	"github.com/pkg/errors"
+
+	"github.com/kopia/kopia/repo/blob"
+	"github.com/kopia/kopia/repo/blob/readonly"
+	"github.com/kopia/kopia/repo/format"
+)
+
+type azPointInTimeStorage struct {
+	azStorage
+
+	pointInTime time.Time
+}
+
+func (az *azPointInTimeStorage) ListBlobs(ctx context.Context, blobIDPrefix blob.ID, cb func(bm blob.Metadata) error) error {
+	var (
+		previousID blob.ID
+		vs         []versionMetadata
+	)
+
+	err := az.listBlobVersions(ctx, blobIDPrefix, func(vm versionMetadata) error {
+		if vm.BlobID != previousID {
+			// different blob, process previous one
+			if v, found := newestAtUnlessDeleted(vs, az.pointInTime); found {
+				if err := cb(v.Metadata); err != nil {
+					return err
+				}
+			}
+
+			previousID = vm.BlobID
+			vs = vs[:0] // reset for next blob
+		}
+
+		vs = append(vs, vm)
+
+		return nil
+	})
+	if err != nil {
+		return errors.Wrapf(err, "could not list blob versions at time %s", az.pointInTime)
+	}
+
+	// process last blob
+	if v, found := newestAtUnlessDeleted(vs, az.pointInTime); found {
+		if err := cb(v.Metadata); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (az *azPointInTimeStorage) GetBlob(ctx context.Context, blobID blob.ID, offset, length int64, output blob.OutputBuffer) error {
+	// getMetadata returns the specific blob version at time t
+	m, err := az.getVersionedMetadata(ctx, blobID)
+	if err != nil {
+		return errors.Wrap(err, "getting metadata")
+	}
+
+	return az.getBlobWithVersion(ctx, blobID, m.Version, offset, length, output)
+}
+
+// newestAtUnlessDeleted returns the last version in the list older than the PIT.
+// Azure sorts in ascending order so return the last element in the list.
+func newestAtUnlessDeleted(vs []versionMetadata, t time.Time) (v versionMetadata, found bool) {
+	vs = getOlderThan(vs, t)
+
+	if len(vs) == 0 {
+		return versionMetadata{}, false
+	}
+
+	v = vs[len(vs)-1]
+
+	return v, !v.IsDeleteMarker
+}
+
+// Removes versions that are newer than t. The filtering is done in place
+// and uses the same slice storage as vs. Assumes entries in vs are in ascending
+// timestamp order (and version order), unlike S3 which assumes descending.
+// Versions in Azure follow the time.RFC3339Nano syntax.
+func getOlderThan(vs []versionMetadata, t time.Time) []versionMetadata {
+	for i := range vs {
+		if vs[i].Timestamp.After(t) {
+			return vs[:i]
+		}
+
+		// The DeleteMarker blob takes the Timestamp of the previous version but has its own Version.
+		// If there was a Kopia Delete Marker (the blob was protected) it will be caught above but if
+		// the container has versioning enabled but no blob retention protection (or the blob was deleted outside
+		// of the protection window) then we need to check the time of the VersionID because there could be a situation
+		// where Azure's DeleteMarker version has Timestamp 2023-10-20 but Version 2023-10-27.
+		if vs[i].IsDeleteMarker {
+			// Assumes the Put and deletion didn't happen within the same second.
```

Comment:
Does this always hold? What are the consequences when the put and delete happen in the same second?

Reply:
Removed this code